src/PackagePrivate/Util/ReplicaMasterAwareRecordIdsAcquirer.php

```diff
@@ -0,0 +1,204 @@
+ */
+class ReplicaMasterAwareRecordIdsAcquirer {
+
+	/**
+	 * @var IDatabase $dbMaster
+	 */
+	private $dbMaster;
+
+	/**
+	 * @var IDatabase $dbReplica
+	 */
+	private $dbReplica;
+
+	/**
+	 * @var string $table
+	 */
+	private $table;
+
+	/**
+	 * @var string $idColumn
+	 */
+	private $idColumn;
+
+	/**
+	 * @var LoggerInterface $logger
+	 */
+	private $logger;
+
+	/**
+	 * @param IDatabase $dbMaster master database to insert non-existing records into
+	 * @param IDatabase $dbReplica replica database to initially query existing records in
+	 * @param string $table the name of the table this acquirer is for
+	 * @param string $idColumn the name of the column that contains the desired ids
+	 * @param LoggerInterface $logger
+	 */
+	public function __construct(
+		IDatabase $dbMaster,
+		IDatabase $dbReplica,
+		$table,
+		$idColumn,
+		LoggerInterface $logger = null
+	) {
+		$this->dbMaster = $dbMaster;
+		$this->dbReplica = $dbReplica;
+		$this->table = $table;
+		$this->idColumn = $idColumn;
+		$this->logger = $logger ?? new NullLogger();
+	}
+
+	/**
+	 * Acquire ids of needed records in the table, inserting non-existing
+	 * ones into master database.
+	 *
+	 * Note 1: this function assumes that all records given in $neededRecords specify
+	 * the same columns. If some records specify less, more or different columns than
+	 * the first one does, the behavior is not defined.
+	 *
+	 * Note 2: this function assumes that all records given in $neededRecords have
+	 * their values as strings. If some values are of different type (e.g. integer ids)
+	 * this can cause infinite loops due to mismatch in identifying records selected in
+	 * database with their corresponding needed records. The first element keys will be
+	 * used as the set of columns to select in database and to provide back in the returned array.
+	 *
+	 * @param array $neededRecords array of records to be looked-up or inserted.
+	 *	Each entry in this array should an associative array of column => value pairs.
+	 *	Example:
+	 *	[
+	 *		[ 'columnA' => 'valueA1', 'columnB' => 'valueB1' ],
+	 *		[ 'columnA' => 'valueA2', 'columnB' => 'valueB2' ],
+	 *		...
+	 *	]
+	 *
+	 * @return array the array of input recrods along with their ids
+	 *	Example:
+	 *	[
+	 *		[ 'columnA' => 'valueA1', 'columnB' => 'valueB1', 'idColumn' => '1' ],
+	 *		[ 'columnA' => 'valueA2', 'columnB' => 'valueB2', 'idColumn' => '2' ],
+	 *		...
+	 *	]
+	 */
+	public function acquireIds( array $neededRecords ) {
+		$existingRecords = $this->findExistingRecords( $this->dbReplica, $neededRecords );
+		$neededRecords = $this->filterNonExistingRecords( $neededRecords, $existingRecords );
+
+		while ( !empty( $neededRecords ) ) {
+			$this->insertNonExistingRecordsIntoMaster( $neededRecords );
+			$existingRecords = array_merge(
+				$existingRecords,
+				$this->findExistingRecords( $this->dbMaster, $neededRecords )
+			);
+			$neededRecords = $this->filterNonExistingRecords( $neededRecords, $existingRecords );
```

Comment:
When would this be non-empty? AFAIS only if something you just inserted into master got deleted… is it really worth guarding just for this specific case? (If you want to be safe in all cases, you need to `SELECT … FOR UPDATE` _all_ record-rows on master)

Reply:
this would be non-empty in the following case:
replica has records 1, 2 but not 3, 4 and 5
master has records 3 and 4 but not 5 (the previous insert will fail due to 3 and 4 duplicatation)
there fore 5 is still needed and to be inserted again into master

Reply:
I thought I added a test case for that one, but the one I have under `testWhenSomeRecordsDoNotExistInReplicaButExistInMaster` tests the case where all $neededRecords spilling here exist in master .. will add another test case for that

Reply:
I changed the existing test case in `testWhenSomeRecordsDoNotExistInReplicaButExistInMaster` so that master have only a subset of those neededRecrods already, in which case the first attempt to save will fail resulting in the $neededRecords to be non-empty on this line.

using infinite loop here might still not be the best option .. we can make it just two steps checks procedural-style .. but I don't know if that would look better so I'd rather not change it until we find a third better improvement

Reply:
Ok, that makes sense… but given we only care for equivalent rows anyway, couldn't we just go with `INSERT IGNORE` and be done?

Reply:
yup we could've but we tried that before and the problem with using IGNORE Is that it will still increment the next primary key per every failing record .. we concluded that we do not want that side-effect in tables with high cardinality as our primary keys will have gaps and might risk reaching much more quickly the maximum value of id column